base/multi.go

```diff
@@ -0,0 +1,104 @@
+package base
+
+import (
+	"database/sql"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+)
+
+type multi struct {
+	sync.Mutex
+	*DebugOutput
+
+	db             *DB
+	id             string
+	isLeader       bool
+	timeoutSeconds int
+	interval       time.Duration
+}
+
+func newMulti(db *DB, debugConfig *ChatDebugOutputConfig) *multi {
+	return &multi{
+		DebugOutput:    NewDebugOutput("Multi", debugConfig),
+		db:             db,
+		timeoutSeconds: 5,
+		interval:       time.Second,
+	}
+}
+
+func (m *multi) Heartbeat(shutdownCh chan struct{}) (err error) {
+	m.id = RandHexString(8)
+	m.Debug("Heartbeat: starting multi coordination heartbeat loop: id: %s", m.id)
+	for {
+		select {
+		case <-time.After(m.interval):
+			m.heartbeat()
+		case <-shutdownCh:
+			m.Debug("Heartbeat: shutdown received, deregistering")
+			m.deregister()
+			return nil
+		}
+	}
+}
+
+func (m *multi) IsLeader() bool {
+	m.Lock()
+	defer m.Unlock()
+	return m.isLeader
+}
+
+func (m *multi) heartbeat() {
+	if err := m.db.RunTxn(func(tx *sql.Tx) error {
+		// update ourselves first
+		_, err := m.db.Exec(`
+			INSERT INTO heartbeats (id, mtime) VALUES (?, NOW(6)) ON DUPLICATE KEY UPDATE mtime=NOW(6)
```

Comment:
I wonder if this and below can be merged with adding a `where` to the `update` clause. So there's gonna be only one row in the table ever, and we only update it if `id` is greater, and `mtime` hasn't expired. Then we know from the result of the insert/update query whether we are the leader.